### ISSUE TITLE
Get the forward/backward of a Dynamo function

### ISSUE THREAD

**Issue by tengyifei** [pytorch/pytorch#140630]
### 🚀 The feature, motivation and pitch

Pitch: Get the forward/backward of a Dynamo function

We (PyTorch/XLA) would like to get a symbolic graph representation (e.g. FX graph) of the forward and backward pass of a torch function, in order to implement a scan operator similar to `jax.lax.scan`. The forward pass returns the output and intermediate activations. The backward pass takes in intermediate activations and gradients w.r.t the output and returns gradients w.r.t the input. The forward of a `scan(fn, ...)` is built using the forward of `fn`, and the backward of a `scan(fn, ...)` is built using the backward of `fn`.

There are a few ways to get a symbolic representation. Since Dynamo has support for higher order operators such as `torch.utils.checkpoint` [1], we would like to explore getting the forward/backward of a `fn` using Dynamo.

I tried a bunch of ways of getting that information without changing PyTorch, and hit challenges. They are documented in this notebook: [2].

[1]: https://dev-discuss.pytorch.org/t/higher-order-operators-2023-10/1565
[2]: https://colab.research.google.com/drive/1qDGY3xPO5JWFJ3jeUHFMD1AnmksDHOOk?usp=sharing

### Alternatives

An alternative to using Dynamo is to use the symbolic tracing facilities in `functorch`. For example, we could use `functorch.compile.aot_function` to get the forward/backward of a `fn`. In our own testing, that works nicely for simple cases involve pure torch numerical operations, but it either errors or ignores `torch.utils.checkpoint` entirely.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @zou3519 @ydwu4 @bdhirsh @yf225 @ezyang

**Comment by shunting314** [comment]
From triage review: 
1. both capture and argument reordering are by design
2. we'll look at the checkpoint problem

**Comment by zou3519** [comment]
We've got a scan coming in the words, do you literally just want a jax.lax.scan for PyTorch?

**Comment by tengyifei** [comment]
What (I think) we want is a Scan node in dynamo so I could lower that into an XLA scan operation via a dynamo compiler backend. Based on my understanding, when encountering a `torch.scan(f, ...)` or any HOP Dynamo will functionalize and lift free variables in `f` and give `f` to the compiler backend. This way I could lower `f` to XLA operations and lower the Scan node to an XLA scan with the aforementioned operations as the combining function.

I looked at `torch.scan` today and it doesn't support autograd. Maybe we need it to support Autograd and this approach will work?

**Comment by ydwu4** [comment]
yeah, we have a few on-going PR stacks for scan and we're actively working on them:
1. jit inductor: https://github.com/pytorch/pytorch/pull/135603
2. Scan autograd PR: https://github.com/pytorch/pytorch/pull/135631
You may patch these PRs to see if it works in your case.

**Comment by rpsilva-aws** [comment]
This would also allows us to leverage this scan for grad accumulation. I see that some of these have been stale. Do we plans to land these in 2.7?
